fendermint/vm/interpreter/src/fvm/query.rs

```diff
@@ -64,24 +66,19 @@ where
                 FvmQueryRet::Call(ret)
             }
             FvmQuery::EstimateGas(mut msg) => {
-                // TODO: Figure out how to relate the token balance to a gas limit.
-                //       Do we look at `state.state_params.base_fee`? What about `gas_premium` and `gas_limit`?
-
-                // XXX: This value is for Filecoin, and it's not even used by the FVM, but at least it should not have a problem with it.
+                // Setting BlockGasLimit as initial limit for gas estimation
                 msg.gas_limit = BLOCK_GAS_LIMIT;
 
-                // TODO: This actually fails if the caller doesn't have enough gas.
-                //       Should we modify the state tree up front to give it more?
-                let apply_ret = state.call(*msg)?;
+                // Populate gas message parameters.
+                // If message fails with BLOCK_GAS_LIMIT as the message gas limit
+                // it means that there is an error in the execution of the message
+                // we could optionally propagate that error if needed as done in Lotus.
+                // (but why would you want to estimate the gas of a message that can't
+                // be executed?)
+                let mut msg = self.estimate_gassed_msg(&state, msg)?;
 
-                let est = GasEstimate {
-                    exit_code: apply_ret.msg_receipt.exit_code,
-                    info: apply_ret
-                        .failure_info
-                        .map(|x| x.to_string())
-                        .unwrap_or_default(),
-                    gas_limit: apply_ret.msg_receipt.gas_used,
-                };
+                // perform a gas search for an accurate value
+                let est = self.gas_search(&state, &mut msg)?;
```

Comment:
I don't think using `?` here is the right call, because further down you return an error if the call fails for any reason other than OUT_OF_GAS, which would kill the proces. 

This is why previously I returned the `ApplyRet` from the `call` and used that to return a `GasEstimate` with both the gas used _and_ the fault code.

Reply:
Now I get how it works :) Fixed in the next commit. Thanks!